tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,438 @@
+    remote op layer.
+  """
+  output_node_names = set([])
+
+  for remote_op in remote_ops_one_layer:
+    for input_node_name in node_name_to_input_names[remote_op]:
+      input_node = node_name_to_node_def[input_node_name]
+
+      # Assumption: graph inputs (placeholders) are loaded before
+      #             executing a graph.
+      if not _is_placeholder_op(input_node):
+        output_node_names.add(input_node_name)
+
+  return output_node_names
+
+
+def _partition_one_subgraph_layer(previous_layers_visited,
+                                  graph_def,
+                                  graph,
+                                  output_names,
+                                  node_name_to_node_def,
+                                  node_name_to_input_names):
+  """Perform a modified BFS for graph partitioning.
+
+  Expand from the outputs, until one of the stopping condition: remote op,
+    placeholder, visited before in this layer, or visited before in the
+    previous layers.
+
+  Args:
+    previous_layers_visited: a set of nodes
+    graph_def: a GraphDef proto
+    graph: a tf.Graph
+    output_names: a set of output names
+    node_name_to_node_def: {node name: a NodeDef proto}
+    node_name_to_input_names: {node name: a list of node input names}
+
+  Returns:
+    An execution spec.
+  """
+  subgraph = graph_pb2.GraphDef()
+  subgraph.versions.CopyFrom(graph_def.versions)
+  subgraph.library.CopyFrom(graph_def.library)
+
+  queue = list(output_names)
+  current_layer_visited = set([])
+  nodes_from_other_layers = set([])
+
+  while queue:
+    current_node_name = queue[0]
+    current_node = node_name_to_node_def[current_node_name]
+    del queue[0]
+
+    if _is_remote_op(current_node) or _is_placeholder_op(current_node):
+      # Remote op or placeholder input will always be prepared.
+      if current_node_name not in current_layer_visited:
+        placeholder_node = _create_placeholder_node_from_existing_node(
+            current_node, graph)
+        subgraph.node.append(placeholder_node)
+
+        current_layer_visited.add(current_node_name)
+    else:
+      # Regular op may be an intermediate node from other graphs and
+      # not prepared, so we need to find them and do something later.
+      if current_node_name in previous_layers_visited:
+        nodes_from_other_layers.add(current_node_name)
+
+      elif current_node_name not in current_layer_visited:
+        subgraph.node.append(current_node)
+
+        current_layer_visited.add(current_node_name)
+        queue.extend(node_name_to_input_names[current_node_name])
+
+  return ExecutionSpec(
+      subgraph=subgraph,
+      input_names=_get_input_names_from_subgraph(subgraph),
+      output_names=set(output_names),
+      is_remote_op=False,
+      body_node_names=_get_body_node_names_from_subgraph(subgraph),
+      nodes_from_other_layers=nodes_from_other_layers)
+
+
+def _handle_nodes_from_other_layers(current_spec,
```

Comment:
What does "handling" nodes from other layers mean? Is there a more specific term we can use to describe this function?

Reply:
Don't have a more specific term in mind :(. We found some nodes coming from other layers, and we'd like to "handle" them.